### PR TITLE
add `--trace-compile-timing` arg to add compile timing comments

### PR DESCRIPTION
**Pull request by IanButterworth**
```
% ./julia --trace-compile=stderr --trace-compile-timing --start=no -e "using ProfileView"
#=   0.0 ms =# precompile(Tuple{typeof(Requires.__init__)})
#=  19.4 ms =# precompile(Tuple{typeof(Requires.isprecompiling)})
#=   0.0 ms =# precompile(Tuple{typeof(Requires.listenpkg), Any, Base.PkgId})
#=   0.0 ms =# precompile(Tuple{typeof(Requires.loaded), Base.PkgId})
#=  13.4 ms =# precompile(Tuple{typeof(Base.haskey), Base.Dict{Base.PkgId, Module}, Base.PkgId})
#=   0.0 ms =# precompile(Tuple{typeof(Requires.callbacks), Base.PkgId})
#=   0.0 ms =# precompile(Tuple{typeof(Libiconv_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Libiconv_jll.find_artifact_dir)})
#=  32.0 ms =# precompile(Tuple{typeof(Artifacts._artifact_str), Module, String, Base.SubString{String}, String, Base.Dict{String, Any}, Base.SHA1, Base.BinaryPlatforms.Platform, Any})
#=   1.5 ms =# precompile(Tuple{typeof(Base.invokelatest), Any})
...
```

### CODE REVIEW

**Comment by tecosaur** [comment]
Love the look of this, the only improvement that comes to mind is rounding the timings printed to 3 sig fig.

File: src/jloptions.c [hunk 1]
```diff
@@ -793,7 +798,7 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
                 jl_errorf("julia: invalid argument to --inline (%s)", optarg);
             }
             break;
-       case opt_polly:
+        case opt_polly:
```
Comment (tecosaur) [inline src/jloptions.c:801]: Unrelated whitespace change?

**Comment by tecosaur** [comment]
With how short and easy to read these changes are, I'd say this LGTM.

File: src/jloptions.c (hunk 1 above)
Comment (IanButterworth) [reply src/jloptions.c:801]: Yeah, flyby formatting fix

**Comment by IanButterworth** [comment]
Updated to 3 d.p.

```
% ./julia -t2 --trace-compile=stderr --trace-compile-timing=yes --start=no
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.12.0-DEV.694 (2024-06-09)
 _/ |\__'_|_|_|\__'_|  |  ib/trace_compile_timing/a16e5a6847 (fork: 2 commits, 0 days)
|__/                   |

(......) pkg> precompile(Tuple{REPL.var"#119#137"{REPL.LineEdit.MIState, REPL.LineEditREPL, REPL.LineEdit.Prompt}}) # 4.538ms
precompile(Tuple{typeof(Base.first), Array{Any, 1}}) # 3.328ms
precompile(Tuple{typeof(Base.put!), Base.Channel{Any}, Function}) # 22.375ms
precompile(Tuple{typeof(Base.take!), Base.Channel{Any}}) # 1.911ms
precompile(Tuple{REPL.var"#120#138"{Module, REPL.LineEditREPL, REPL.LineEdit.Prompt}, REPL.LineEdit.MIState, Any}) # 5.024ms
precompile(Tuple{REPL.var"#121#139"{REPL.LineEdit.MIState, REPL.LineEdit.Prompt}}) # 2.808ms
```

**Comment by IanButterworth** [comment]
Just noting that this is quite nice in conjunction with `@time_imports` for attributing where the compilation time is spent

File: src/gf.c [hunk 2]
```diff
@@ -2478,7 +2478,12 @@ static void record_precompile_statement(jl_method_instance_t *mi)
     if (!jl_has_free_typevars(mi->specTypes)) {
         jl_printf(s_precompile, "precompile(");
         jl_static_show(s_precompile, mi->specTypes);
-        jl_printf(s_precompile, ")\n");
+        if (jl_options.trace_compile_timing) {
+            jl_printf(s_precompile, ") # %.3fms\n", compilation_time / 1e6);
```
Comment (tecosaur) [inline src/gf.c:2482]: Nit/extra-minor tweak: how about "%.3g`?

Comment (IanButterworth) [reply src/gf.c:2482]: We tend to use fixed d.p. in timing rather than s.f.  Like in `@time` and `precompile(timing=true)`

Comment (tecosaur) [reply src/gf.c:2482]: Cool, I was just thinking might as well not show more information is helpful/reliable and that 2-3 sig fig should be plenty from that perspective.

Comment (IanButterworth) [reply src/gf.c:2482]: I think d.p. makes more sense for quick visual scanning because `12.345` looks bigger than `1.234` 
Compared to s.f. `12.3` vs. `1.23`

**Comment by IanButterworth** [comment]
Looking at https://github.com/JuliaLang/julia/pull/51106 it's probably better to implement this as `--trace-compile-timing` rather than require the value `--trace-compile-timing=yes`

File: src/gf.c (hunk 2 above)
Comment (tecosaur) [reply src/gf.c:2482]: Perhaps, but then you've got `ms` and `us` :shrug:.

If we're going with fixed dp though, I'd suggest that `%.1-2f` would be better that `%.3f`. It's even easier to visually scan `12.3` vs. `1.2` (given system variation, I'm not sure when 2 let alone 3 dp would actually be significant).

**Comment by IanButterworth** [comment]
Now implemented via simply `--trace-compile-timing` (no `=yes`)

Also to make the information easier to compare, the info has been moved to before the statement, in an inline comment so that the raw output could still be loaded in a `.jl`.

Also the time is printed fixed width left padded, to align the decimal in the same way that our other timing tooling does
```
% ./julia -t2 --trace-compile=stderr --trace-compile-timing=yes --start=no -e "using ProfileView"
#=   0.0 ms =# precompile(Tuple{typeof(Requires.__init__)})
#=  19.4 ms =# precompile(Tuple{typeof(Requires.isprecompiling)})
#=   0.0 ms =# precompile(Tuple{typeof(Requires.listenpkg), Any, Base.PkgId})
#=   0.0 ms =# precompile(Tuple{typeof(Requires.loaded), Base.PkgId})
#=  13.4 ms =# precompile(Tuple{typeof(Base.haskey), Base.Dict{Base.PkgId, Module}, Base.PkgId})
#=   0.0 ms =# precompile(Tuple{typeof(Requires.callbacks), Base.PkgId})
#=   0.0 ms =# precompile(Tuple{typeof(Libiconv_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Libiconv_jll.find_artifact_dir)})
#=  32.0 ms =# precompile(Tuple{typeof(Artifacts._artifact_str), Module, String, Base.SubString{String}, String, Base.Dict{String, Any}, Base.SHA1, Base.BinaryPlatforms.Platform, Any})
#=   1.5 ms =# precompile(Tuple{typeof(Base.invokelatest), Any})
#=   0.0 ms =# precompile(Tuple{typeof(JLLWrappers.get_julia_libpaths)})
#=   0.0 ms =# precompile(Tuple{typeof(Libffi_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Libffi_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(XML2_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(XML2_jll.find_artifact_dir)})
#=   5.7 ms =# precompile(Tuple{typeof(Base.issorted), Array{String, 1}, Base.Order.ReverseOrdering{Base.Order.ForwardOrdering}})
#=   0.0 ms =# precompile(Tuple{typeof(Gettext_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Gettext_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Glib_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Glib_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(LLVMOpenMP_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(LLVMOpenMP_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Pixman_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Pixman_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(libpng_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(libpng_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Bzip2_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Bzip2_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(FreeType2_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(FreeType2_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Expat_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Expat_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Fontconfig_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Fontconfig_jll.find_artifact_dir)})
#=   4.1 ms =# precompile(Tuple{typeof(Base.get!), Base.EnvDict, String, String})
#=   2.1 ms =# precompile(Tuple{typeof(Base.getindex), Base.EnvDict, String})
#=   0.0 ms =# precompile(Tuple{typeof(Cairo_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Cairo_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(FriBidi_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(FriBidi_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Graphite2_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Graphite2_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(HarfBuzz_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(HarfBuzz_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Pango_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Pango_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(JpegTurbo_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(JpegTurbo_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(LERC_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(LERC_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(XZ_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(XZ_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Zstd_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Zstd_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Libtiff_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Libtiff_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(gdk_pixbuf_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(gdk_pixbuf_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Libepoxy_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Libepoxy_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(ATK_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(ATK_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(iso_codes_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(iso_codes_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(GTK3_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(GTK3_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Xorg_xkeyboard_config_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Xorg_xkeyboard_config_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(hicolor_icon_theme_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(hicolor_icon_theme_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(adwaita_icon_theme_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(adwaita_icon_theme_jll.find_artifact_dir)})
#=   0.0 ms =# precompile(Tuple{typeof(Librsvg_jll.__init__)})
#=   0.0 ms =# precompile(Tuple{typeof(Librsvg_jll.find_artifact_dir)})
#=   7.5 ms =# precompile(Tuple{Gtk.var"#query_pixbuf_loaders#327"{Gtk.var"#query_pixbuf_loaders#322#328"}, String})
#=   0.0 ms =# precompile(Tuple{typeof(gdk_pixbuf_jll.gdk_pixbuf_query_loaders)})
#=   0.0 ms =# precompile(Tuple{gdk_pixbuf_jll.var"##gdk_pixbuf_query_loaders#4", Bool, Bool, typeof(gdk_pixbuf_jll.gdk_pixbuf_query_loaders)})
#=   4.1 ms =# precompile(Tuple{typeof(Base.copy), Base.EnvDict})
#=   1.4 ms =# precompile(Tuple{typeof(Base.getindex), Base.RefValue{String}})
#=   0.0 ms =# precompile(Tuple{typeof(JLLWrappers.adjust_ENV!), Base.Dict{K, V} where V where K, String, String, Bool, Bool})
#=   1.6 ms =# precompile(Tuple{typeof(Base.get), Base.Dict{String, String}, String, String})
#=   1.0 ms =# precompile(Tuple{typeof(Base.string), String, Char, Vararg{Union{Char, String, Symbol}}})
#=   1.6 ms =# precompile(Tuple{typeof(Base.getindex), Base.Dict{String, String}, String})
#=   1.6 ms =# precompile(Tuple{Type{Base.Cmd}, Array{String, 1}})
#=   1.2 ms =# precompile(Tuple{Type{NamedTuple{(:env,), T} where T<:Tuple}, Tuple{Base.Dict{String, String}}})
#=  28.5 ms =# precompile(Tuple{typeof(Base.addenv), Base.Cmd, Pair{String, String}})
#=   2.0 ms =# precompile(Tuple{typeof(Base.read), Base.PipeEndpoint})
#=   1.7 ms =# precompile(Tuple{typeof(Base.write), Base.IOStream, Base.SubString{String}})
#=   0.0 ms =# precompile(Tuple{typeof(Librsvg_jll.is_available)})
#=   2.1 ms =# precompile(Tuple{typeof(Core.kwcall), NamedTuple{(:extra_env,), Tuple{Array{Pair{String, String}, 1}}}, Gtk.var"#query_pixbuf_loaders#327"{Gtk.var"#query_pixbuf_loaders#322#328"}, String})
#=   6.5 ms =# precompile(Tuple{typeof(Base.addenv), Base.Cmd, Pair{String, String}, Pair{String, String}})
#=   1.6 ms =# precompile(Tuple{typeof(Requires.isprecompiling)})
#=   0.0 ms =# precompile(Tuple{typeof(Requires.loadpkg), Base.PkgId})
#=   2.5 ms =# precompile(Tuple{typeof(Base.first), Array{Any, 1}})
#=   1.3 ms =# precompile(Tuple{Gtk.GLib.var"#100#101"})
```

**Comment by IanButterworth** [comment]
Now only printing as a comment when sent to file
```
% ./julia --trace-compile=stderr --trace-compile-timing --start=no -e "using ProfileView"
  36.2 ms precompile(Tuple{typeof(Base.setindex!), Base.EnvDict, String, String})
  10.3 ms precompile(Tuple{typeof(Base.getindex), Base.EnvDict, String})
   0.0 ms precompile(Tuple{typeof(Requires.__init__)})
   7.0 ms precompile(Tuple{typeof(Requires.isprecompiling)})
   0.0 ms precompile(Tuple{typeof(Requires.listenpkg), Any, Base.PkgId})
   0.0 ms precompile(Tuple{typeof(Requires.loaded), Base.PkgId})
   4.1 ms precompile(Tuple{typeof(Base.haskey), Base.Dict{Base.PkgId, Module}, Base.PkgId})
   0.0 ms precompile(Tuple{typeof(Requires.callbacks), Base.PkgId})
   0.0 ms precompile(Tuple{typeof(Libiconv_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Libiconv_jll.find_artifact_dir)})
  48.6 ms precompile(Tuple{typeof(Artifacts._artifact_str), Module, String, Base.SubString{String}, String, Base.Dict{String, Any}, Base.SHA1, Base.BinaryPlatforms.Platform, Any})
   2.6 ms precompile(Tuple{typeof(Base.invokelatest), Any})
   0.0 ms precompile(Tuple{typeof(JLLWrappers.get_julia_libpaths)})
   0.0 ms precompile(Tuple{typeof(Libffi_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Libffi_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(XML2_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(XML2_jll.find_artifact_dir)})
  10.2 ms precompile(Tuple{typeof(Base.issorted), Array{String, 1}, Base.Order.ReverseOrdering{Base.Order.ForwardOrdering}})
   0.0 ms precompile(Tuple{typeof(Gettext_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Gettext_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Glib_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Glib_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(LLVMOpenMP_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(LLVMOpenMP_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Pixman_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Pixman_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(libpng_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(libpng_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Bzip2_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Bzip2_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(FreeType2_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(FreeType2_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Expat_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Expat_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Fontconfig_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Fontconfig_jll.find_artifact_dir)})
   7.3 ms precompile(Tuple{typeof(Base.get!), Base.EnvDict, String, String})
   0.0 ms precompile(Tuple{typeof(Cairo_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Cairo_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(FriBidi_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(FriBidi_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Graphite2_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Graphite2_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(HarfBuzz_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(HarfBuzz_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Pango_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Pango_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(JpegTurbo_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(JpegTurbo_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(LERC_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(LERC_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(XZ_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(XZ_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Zstd_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Zstd_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Libtiff_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Libtiff_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(gdk_pixbuf_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(gdk_pixbuf_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Libepoxy_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Libepoxy_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(ATK_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(ATK_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(iso_codes_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(iso_codes_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(GTK3_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(GTK3_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Xorg_xkeyboard_config_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Xorg_xkeyboard_config_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(hicolor_icon_theme_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(hicolor_icon_theme_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(adwaita_icon_theme_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(adwaita_icon_theme_jll.find_artifact_dir)})
   0.0 ms precompile(Tuple{typeof(Librsvg_jll.__init__)})
   0.0 ms precompile(Tuple{typeof(Librsvg_jll.find_artifact_dir)})
  15.6 ms precompile(Tuple{Gtk.var"#query_pixbuf_loaders#327"{Gtk.var"#query_pixbuf_loaders#322#328"}, String})
   0.0 ms precompile(Tuple{typeof(gdk_pixbuf_jll.gdk_pixbuf_query_loaders)})
   0.0 ms precompile(Tuple{gdk_pixbuf_jll.var"##gdk_pixbuf_query_loaders#4", Bool, Bool, typeof(gdk_pixbuf_jll.gdk_pixbuf_query_loaders)})
   5.7 ms precompile(Tuple{typeof(Base.copy), Base.EnvDict})
   2.5 ms precompile(Tuple{typeof(Base.getindex), Base.RefValue{String}})
   0.0 ms precompile(Tuple{typeof(JLLWrappers.adjust_ENV!), Base.Dict{K, V} where V where K, String, String, Bool, Bool})
   3.1 ms precompile(Tuple{typeof(Base.get), Base.Dict{String, String}, String, String})
   2.0 ms precompile(Tuple{typeof(Base.string), String, Char, Vararg{Union{Char, String, Symbol}}})
   3.2 ms precompile(Tuple{typeof(Base.getindex), Base.Dict{String, String}, String})
   3.2 ms precompile(Tuple{Type{Base.Cmd}, Array{String, 1}})
   2.5 ms precompile(Tuple{Type{NamedTuple{(:env,), T} where T<:Tuple}, Tuple{Base.Dict{String, String}}})
  57.7 ms precompile(Tuple{typeof(Base.addenv), Base.Cmd, Pair{String, String}})
   4.3 ms precompile(Tuple{typeof(Base.read), Base.PipeEndpoint})
   3.2 ms precompile(Tuple{typeof(Base.write), Base.IOStream, Base.SubString{String}})
   0.0 ms precompile(Tuple{typeof(Librsvg_jll.is_available)})
   4.2 ms precompile(Tuple{typeof(Core.kwcall), NamedTuple{(:extra_env,), Tuple{Array{Pair{String, String}, 1}}}, Gtk.var"#query_pixbuf_loaders#327"{Gtk.var"#query_pixbuf_loaders#322#328"}, String})
  13.5 ms precompile(Tuple{typeof(Base.addenv), Base.Cmd, Pair{String, String}, Pair{String, String}})
   3.3 ms precompile(Tuple{typeof(Requires.isprecompiling)})
   0.0 ms precompile(Tuple{typeof(Requires.loadpkg), Base.PkgId})
   3.8 ms precompile(Tuple{typeof(Base.first), Array{Any, 1}})
   2.2 ms precompile(Tuple{Gtk.GLib.var"#100#101"})
   ```

**Comment by KristofferC** [comment]
I would probably always write the comments since it is emitting Julia code after all and it is often nice to copy paste this even from a terminal output.

**Review by charleskawczynski — COMMENTED** [review]
It’d be pretty nice if we could add an option that would sort this list by time

**Comment by IanButterworth** [comment]
These are output as compilation happens, nothing is stored. I believe SnoopCompile already does that kind of thing https://timholy.github.io/SnoopCompile.jl/dev/

**Comment by KristofferC** [comment]
Having some test code that at least exercises the new command line flag would be good.

File: test/cmdlineargs.jl [hunk 3]
```diff
@@ -793,6 +793,17 @@ let exename = `$(Base.julia_cmd()) --startup-file=no --color=no`
...
 
+    # --trace-compile-timing
+    let
+        io = IOBuffer()
+        v = writereadpipeline(
+            "foo(x) = begin Base.Experimental.@force_compile; x; end; foo(1)",
+            `$exename --trace-compile=stderr --trace-compile-timing -i`,
+            stderr=io)
+        _stderr = String(take!(io))
+        @test occursin(" ms =# precompile(Tuple{typeof(Main.foo), Int64})", _stderr)
```
Comment (IanButterworth) [inline test/cmdlineargs.jl:804]:
```suggestion
        @test occursin(" ms =# precompile(Tuple{typeof(Main.foo), Int", _stderr)
```

File: src/gf.c [hunk 4]
```diff
@@ -2650,13 +2652,15 @@ jl_code_instance_t *jl_compile_method_internal(jl_method_instance_t *mi, size_t
         }
 
         JL_GC_PUSH1(&codeinst);
+        double compile_time = jl_hrtime();
         int did_compile = jl_compile_codeinst(codeinst);
+        compile_time = jl_hrtime() - compile_time;
```
Comment (NHDaly) [inline src/gf.c:2657]: @IanButterworth: Can you help me understand why we have so many different variants of measuring this metric (compilation time)?

I was surprised to see we had to add a new start/stop measurement here. We already have a global metric for compilation time, which has timings around inference and llvm codegen. Why didn't we add to the per-method-instance compilation time counters from those places?

Or, conversely, if this spot can correctly capture all compilation time for a given method instance, can we delete those other spots and incremental global compilation time here?

- https://github.com/JuliaLang/julia/blob/fe864374d50b121aeb3ce5473c206338b21df5e5/src/gf.c#L4390-L4396
- https://github.com/JuliaLang/julia/blob/a98f3713cb9aea3225cdb0e3ec0dd719f556a974/src/aotcompile.cpp#L571-L572
- https://github.com/JuliaLang/julia/blob/a98f3713cb9aea3225cdb0e3ec0dd719f556a974/src/jitlayers.cpp#L578-L579
- ... etc

What's the difference between these approaches? Can we unify them?
Thanks!

**Review by NHDaly — COMMENTED** [review]
Love this feature! 😍 Thanks Ian! :)

We're also using this in the compilation caching effort that @kpamnany and @Drvi are implementing for RelationalAI, to prioritize which methodinstances we precompile for our customers based on their compile time. Thanks for this, Ian! :)

File: src/gf.c (hunk 4 above)
Comment (IanButterworth) [reply src/gf.c:2657]: Maybe that can be reused, yeah. The timer added here seems lower cost than atomic fetches etc. but perhaps it's overcounting.
